pkg/volume/gce_pd/gce_util.go

```diff
@@ -143,84 +153,105 @@ func verifyAttached(pd *gcePersistentDisk, sdBeforeSet util.StringSet, gce cloud
+	glog.V(5).Infof("detachDiskAndVerify for pd %q.", pd.pdName)
 	defer util.HandleCrash()
 
-	// Setting bufferSize to 0 so that when senders send, they are blocked until we recieve. This avoids the need to have a separate exit check.
+	// Start operation, so that other threads can wait on this detach operation.
+	// Set bufferSize to 0 so senders are blocked on send until we recieve.
 	ch, err := detachCleanupManager.Start(pd.pdName, 0 /* bufferSize */)
 	if err != nil {
 		glog.Errorf("Error adding %q to detachCleanupManager: %v", pd.pdName, err)
 		return
 	}
+
 	defer detachCleanupManager.Close(pd.pdName)
 
-	devicePaths := getDiskByIdPaths(pd)
-	for numRetries := 0; numRetries < maxRetries; numRetries++ {
-		for numChecks := 0; numChecks < maxChecks; numChecks++ {
+	defer func() {
+		// Unblock any callers that have been waiting for this detach routine to complete.
+		for {
 			select {
 			case <-ch:
-				glog.Warningf("Terminating GCE PD %q detach verification. Another attach/detach call was made for this PD.", pd.pdName)
-				return
+				glog.V(5).Infof("detachDiskAndVerify for pd %q clearing chan.", pd.pdName)
 			default:
-				allPathsRemoved := true
-				for _, path := range devicePaths {
-					if err := udevadmChangeToDrive(path); err != nil {
-						// udevadm errors should not block disk detachment, log and continue
-						glog.Errorf("%v", err)
-					}
-					if exists, err := pathExists(path); err != nil {
-						glog.Errorf("Error check path: %v", err)
-						return
-					} else {
-						allPathsRemoved = allPathsRemoved && !exists
-					}
-				}
-				if allPathsRemoved {
-					// All paths to the PD have been succefully removed
-					glog.V(5).Infof("Succesfully detached GCE PD %q.", pd.pdName)
-					return
-				}
+				glog.V(5).Infof("detachDiskAndVerify for pd %q done clearing chans.", pd.pdName)
+				return
+			}
+		}
+	}()
 
-				// Sleep then check again
-				glog.V(5).Infof("Waiting for GCE PD %q to detach.", pd.pdName)
-				time.Sleep(checkSleepDuration)
+	devicePaths := getDiskByIdPaths(pd)
+	var gce cloudprovider.Interface
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if gce == nil {
+			var err error
+			gce, err = cloudprovider.GetCloudProvider("gce", nil)
+			if err != nil {
+				// Retry on error. See issue #11321
+				glog.Errorf("Error getting GCECloudProvider while detaching PD %q: %v", pd.pdName, err)
+				time.Sleep(errorSleepDuration)
+				continue
 			}
 		}
 
-		// Try detaching disk again
-		glog.Warningf("Timed out waiting for GCE PD %q to detach. Retrying detach.", pd.pdName)
+		if numRetries > 0 {
+			glog.Warningf("Timed out waiting for GCE PD %q to detach. Retrying detach.", pd.pdName)
+		}
+
 		if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
-			glog.Errorf("Error on retry detach PD %q: %v", pd.pdName, err)
-			return
+			// Retry on error. See issue #11321
+			glog.Errorf("Error detaching PD %q: %v", pd.pdName, err)
```

Comment:
shouldn't we add "sleep X and continue" here?
IIUC, if AttachDisk returns an error, there is no chance that disk will be attached at some point.